components/formatted_markdown_message.jsx

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {FormattedHTMLMessage, injectIntl, intlShape} from 'react-intl';
+import PropTypes from 'prop-types';
+import marked from 'marked';
+
+const UNESCAPED_TAGS = ['BR'];
+const TARGET_BLANK_URL_PREFIX = '!';
+
+class FormattedMarkdownMessage extends React.PureComponent {
+    constructor(props) {
+        super(props);
+        this.renderer = new marked.Renderer();
+        this.renderer.link = FormattedMarkdownMessage.linkF;
+    }
+
+    static linkF(href, title, text) {
+        if (href[0] === TARGET_BLANK_URL_PREFIX) {
+            return `<a href="${href.substring(1, href.length)}" target="_blank">${text}</a>`;
+        }
+        return `<a href="${href}">${text}</a>`;
+    }
+
+    static get propTypes() {
+        return {
+            intl: intlShape.isRequired,
+            id: PropTypes.string.isRequired,
+            defaultMessage: PropTypes.string.isRequired,
+            values: PropTypes.object,
+        };
+    }
+
+    static escapeHtml(str) {
+        var div = document.createElement('div');
+        div.appendChild(document.createTextNode(str));
+        return div.innerHTML;
+    }
+
+    static sanitizer(p1) {
+        const parser = new DOMParser();
+        const htmlDoc = parser.parseFromString(p1, 'text/html');
+        const childNodes = htmlDoc.getElementsByTagName('body')[0].childNodes;
+        if (childNodes.length > 0) {
+            const {tagName} = childNodes[0];
+            if (UNESCAPED_TAGS.includes(tagName)) {
+                return p1;
+            }
+        }
+        return FormattedMarkdownMessage.escapeHtml(p1);
+    }
+
+    render() {
+        const origMsg = this.props.intl.formatMessage({
+            id: this.props.id,
+            defaultMessage: this.props.defaultMessage,
+        }, this.props.values);
+        return (
+            <FormattedHTMLMessage
```

Comment:
Is there a way to do this without using a fake FormattedHTMLMessage? If you define a static `contextTypes` field like [this](https://github.com/mattermost/mattermost-mobile/blob/master/app/screens/clock_display/clock_display_base.js#L19), you can get access to the `this.context.intl.formatMessage(message, values)` that `FormattedHTMLMessage` uses and format it by hand.

Reply:
I totally agree with this comment, probably would be cleaner to use intl.formatMessage

Reply:
I'm not totally following. I think a few things are confusing me. One being that the component is using `formatMessage` a few lines up already. Not sure if that was different than the proposal? The other is that I just wanted to clarify that it's not so much a "fake" component. It's using the real component but it's just skipping the first part of the algorithm (checking for the message in id) and going straight to the `defaultMessage` property. With the knowledge of those things could you elaborate on the proposal? I agree that what is currently here looks a little ugly because of the unmatchable id.

Reply:
Oh, I didn't realize that we were calling formatMessage above. Then I'd definitely like to get rid of this component then because we're formatting the message ourselves before using a `FormattedHTMLMessage` as a more complex version of a span with a `dangerouslySetInnerHTML` prop like
```
return <span dangerouslySetInnerHTML={{__html: marked(...)}}/>;
```

Reply:
I like this proposal of using the span directly. I'll try it out.